File: lua/lush/process.lua
package.preload["lush.process"] = function()
  local scheduler = require("lush.scheduler")
  local process = {}

  function process.capture(command)
    return _lush_process_capture(command)
  end

  function process.spawn(command)
    return _lush_process_spawn(command)
  end

  function process.watch(command, interval_seconds, callback, opts)
    opts = opts or {}
    local queue_policy = opts.queue_policy or "latest" -- "latest" | "drop"
    local in_flight = false
    local pending = false

    local function run()
      if in_flight then
        if queue_policy == "latest" then
          pending = true
        end
        return
      end

      in_flight = true
      _lush_process_capture_async(command, function(output)
        in_flight = false

        local ok, err = pcall(callback, output)
        if not ok then
          print("lush.process.watch callback failed: " .. tostring(err))
        end

        if pending then
          pending = false
          run()
        end
      end)
    end

    return scheduler.every(interval_seconds, run)
  end

  return process
end

File: src/config/runtime_api/process.rs
use std::process::Command;

use glib::MainContext;
use mlua::{Function, Lua, Result};

pub fn install(lua: &Lua) -> Result<()> {
    lua.globals().set(
        "_lush_process_capture",
        lua.create_function(move |_, command: String| {
            let output = Command::new("sh").arg("-c").arg(&command).output();
            let text = match output {
                Ok(out) => String::from_utf8_lossy(&out.stdout).trim().to_string(),
                Err(_) => String::new(),
            };
            Ok(text)
        })?,
    )?;

    lua.globals().set(
        "_lush_process_spawn",
        lua.create_function(move |_, command: String| {
            let ok = Command::new("sh").arg("-c").arg(&command).spawn().is_ok();
            Ok(ok)
        })?,
    )?;

    lua.globals().set(
        "_lush_process_capture_async",
        lua.create_function(move |_, (command, callback): (String, Function)| {
            let (tx, rx) = async_channel::bounded::<String>(1);
            std::thread::spawn(move || {
                let output = Command::new("sh").arg("-c").arg(&command).output();
                let text = match output {
                    Ok(out) => String::from_utf8_lossy(&out.stdout).trim().to_string(),
                    Err(_) => String::new(),
                };
                let _ = tx.send_blocking(text);
            });

            MainContext::default().spawn_local(async move {
                if let Ok(text) = rx.recv().await
                    && let Err(err) = callback.call::<()>(text)
                {
                    log::warn!("lush process async callback failed: {}", err);
                }
            });
            Ok(())
        })?,
    )?;

    Ok(())
}
